src/loaders/zarr_utils/ChunkPrefetchIterator.ts

```diff
@@ -0,0 +1,119 @@
+import { TCZYX } from "./types";
+
+type TZYX = [number, number, number, number];
+
+/**
+ * Directions in which to move outward from the loaded set of chunks while prefetching.
+ *
+ * Ordered in pairs of opposite directions both because that's a sensible order in which to prefetch for our purposes,
+ * and because it lets us treat the least significant bit as the sign. So `direction >> 1` gives the index of the
+ * direction in TZYX-ordered arrays, and `direction & 1` gives the sign of the direction (e.g. positive vs negative Z).
+ *
+ * Note that this enum is mostly for documentation, and its variants are never explicitly used.
+ */
+const enum PrefetchDirection {
+  T_MINUS = 0,
+  T_PLUS = 1,
+
+  Z_MINUS = 2,
+  Z_PLUS = 3,
+
+  Y_MINUS = 4,
+  Y_PLUS = 5,
+
+  X_MINUS = 6,
+  X_PLUS = 7,
+}
+
+type PrefetchDirectionState = {
+  direction: PrefetchDirection;
+  chunks: TCZYX<number>[];
+  start: number;
+  end: number;
+};
+
+const skipC = (idx: number): number => idx + Number(idx !== 0);
+const directionToIndex = (dir: PrefetchDirection): number => skipC(dir >> 1);
+
+/**
+ * Since the user is most likely to want nearby data (in space or time) first, we should prefetch those chunks first.
+ *
+ * Given a list of just-loaded chunks and some bounds, `ChunkPrefetchIterator` iterates evenly outwards in T/Z/Y/X.
+ */
+// NOTE: Assumes `chunks` form a rectangular prism! Will create gaps otherwise! (in practice they always should)
+export default class ChunkPrefetchIterator {
+  directions: PrefetchDirectionState[];
+
+  constructor(chunks: TCZYX<number>[], tzyxMaxOffset: TZYX, tzyxChunkSize: TZYX) {
+    // Get max and min chunk coordinates for T/Z/Y/X
+    const extrema = [Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity];
+
+    for (const chunk of chunks) {
+      for (let j = 0; j < 4; j++) {
+        const val = chunk[skipC(j)];
+        const extremaIdx = j << 1;
```

Comment:
Nit: More readable to as `j * 2`?

Reply:
agree

Reply:
Nit: I get what this is doing, but this code is a bit hard to parse when reading it.   I feel like it might be even more readable if the loop were unrolled and/or the extrema array were more structured, since the dims are always t,z,y,x.   It's a 4D bounding box!  The skipC function is hard to understand too.
pseudocode : 
for each chunk:  
   update min/max of T value
   update min/max of Z value
   update min/max of Y value
   update min/max of X value